samcli/cli/cli_config_file.py

```diff
@@ -0,0 +1,152 @@
+"""
+CLI configuration decorator to use TOML configuration files for click commands.
+"""
+
+## This section contains code copied and modified from [click_config_file][https://github.com/phha/click_config_file/blob/master/click_config_file.py]
+## SPDX-License-Identifier: MIT
+
+import functools
+import os
+import logging
+
+import click
+import toml
+
+__all__ = ("TomlProvider", "configuration_option")
+
+LOG = logging.getLogger(__name__)
+DEFAULT_CONFIG_FILE_NAME = "sam-config"
+DEFAULT_IDENTIFER = "default"
+
+
+class TomlProvider:
+    """
+    A parser for toml configuration files
+    :param cmd: sam command name as defined by click
+    :param section: section defined in the configuration file nested within `cmd`
+    """
+
+    def __init__(self, cmd=None, section=None):
+        self.cmd = cmd
+        self.section = section
+
+    def __call__(self, file_path, identifier, cmd_name):
+        """
+        :param file_path: The path to the configuration file
+        :param identifier: The name of the sectional identifier within configuration file.
+        :param cmd_name: sam command name as defined by click
+        :returns dictionary containing the configuration parameters under specified identifier
+        """
+        config = toml.load(file_path)
+        resolved_config = {}
+        if self.section:
+            try:
+                resolved_config = self._get_identifier(config, identifier)[self.cmd][self.section]
+            except KeyError:
+                LOG.debug(
+                    "Error reading configuration file at %s with identifier %s, command %s, section %s",
+                    file_path,
+                    identifier,
+                    self.cmd,
+                    self.section,
+                )
+        return resolved_config
+
+    def _get_identifier(self, config, identifier):
+        """
+
+        :param config: loaded TOML configuration file into dictionary representation
+        :param identifier: top level section defined within TOML configuration file
+        :return:
+        """
+        return config.get(identifier)
+
+
+def configuration_callback(cmd_name, option_name, identifier_name, saved_callback, provider, ctx, param, value):
+    """
+    Callback for reading the config file.
+
+    Also takes care of calling user specified custom callback afterwards.
+
+    :param cmd_name: `sam` command name derived from click.
+    :param option_name: The name of the option. This is used for error messages.
+    :param identifier_name: `top` level section within configuration file
+    :param saved_callback: User-specified callback to be called later.
+    :param provider:  A callable that parses the configuration file and returns a dictionary
+        of the configuration parameters. Will be called as
+        `provider(file_path, identifier, cmd_name)`.
+    :param ctx: Click context
+    :param param: Click parameter
+    :param value: Specified value for identifier
```

Comment:
can you provide a better variable name than `value`? It is too vague.

Reply:
ctx, param, value are used in click callbacks, so re-using them.